### ISSUE TITLE
Support for LLaMA or Baichuan models

### ISSUE THREAD

**Issue by prajwalganugula** [asahi417/lm-vocab-trimmer#3]
Hi, I was wondering if this method can be used for trimming large vocabulary in LLMs. Can vocab trimmer be extended to LLMs?

**Comment by asahi417** [comment]
It should be possible to apply vocabulary trimming (VT) to mono-lingual LMs, but it would have little affect on the model size. Mainly because for monolingual LLMs, the ratio of embedding matrix to the whole parameters is very little, and in fact their vocabulary is already quite small (eg. LLaMA has 32k tokens). 

The reason why VT is effective for small multilingual LMs (XLM, mT5, mBART), is because the ratio of embedding matrix to the whole model parameter is very dominant (eg. the ratio is 80% for mT5 small), and their vocabulary is extremely large (eg. mT5 has 250k tokens).

Although, it sounds an interesting direction to apply VT to monolingual LMs! I might work on it if I would be more convinced and had some spare time on it!

**Comment by prajwalganugula** [comment]
I agree it has a little effect on LLaMA. But for bilingual LLMs such as Baichuan-2 has 120k tokens. Many multi-lingual LLMs are emerging and this technique would be very useful to make them monolingual and users can finetune in their target languages without worrying about other languages. 

I am currently trying to to extend the library to Baichuan-2. Facing many issues and debugging them... I will release it here after I finish that. May be we can extend this to any LLM on hugging face using Auto Casual LM inference.

**Comment by kalradivyanshu** [comment]
Hey @Tyler-Durden-official were you able to figure this out? I am interested in using this to reduce newer models like phi-3/LLaMA3.

**Comment by urroxyz** [comment]
Support for MistralForCausalLM would also be great!

**Comment by Said-Akbar** [comment]
I was searching for a similar technique and found out this article but I don't have full access since it is paid.
It reduces llama3.1 8B fp16 from 16GB to 14.5GB (~10% reduction). When converted to 4bits this model reduces from 5.8GB to 4.2GB (-72% reduction in size) which is impression.

Article: https://kaitchup.substack.com/p/introducing-minivoc-faster-and-memory-llms
Model: https://huggingface.co/kaitchup/Llama-3.1-8B-Minivoc-32k-v0.1a
